doc/man/hal-is-caller-locked-out.1.in
.\" 
.\" hal-is-caller-locked-out manual page.
.\" Copyright (C) 2007 David Zeuthen <david@fubar.dk>
.\"
.TH HAL-IS-CALLER-LOCKED-OUT 1
.SH NAME
hal-is-caller-locked-out \- determine if a caller is locked out
.SH SYNOPSIS
.PP
.B hal-is-caller-locked-out
[options]

.SH DESCRIPTION

\fIhal-is-caller-locked-out\fP determines if a specific caller is
locked out of a specific D-Bus interface on a specific device. For
more information about both the big picture and specific
.B HAL
properties, refer to the \fIHAL spec\fP which can be found in
.I "@docdir@/spec/hal-spec.html"
depending on the distribution.

.SH OPTIONS
The following options are supported:
.TP
.I "--udi"
The UDI (\fIUnique Device Identifier\fP) of the device object.
.TP
.I "--interface"
The name of the interface to check for.
.TP
.I "--caller"
The unique D-Bus name on the system bus of the caller.
.TP
.I "--help"
Print out usage.
.TP
.I "--version"
Print the version.

.SH RETURN VALUE
.PP
If the given process is locked out or an error occurs, this program
exits with exit code 1. Otherwise it exits with exit code 0. Note that
only the super user (root) or other privileged users can use this
tool.

.SH BUGS
.PP
Please send bug reports to either the distribution or the HAL
mailing list, see 
.I "http://lists.freedesktop.org/mailman/listinfo/hal"
on how to subscribe.

.SH SEE ALSO
.PP
\&\fIhald\fR\|(8), 
\&\fIlshal\fR\|(1), 
\&\fIhal-get-property\fR\|(1),
\&\fIhal-set-property\fR\|(1),
\&\fIhal-find-by-property\fR\|(1),
\&\fIhal-find-by-capability\fR\|(1)

.SH AUTHOR
Written by David Zeuthen <david@fubar.dk> with a lot of help from many
others.



NEWS
==========
HAL 0.5.10 ""
==========

Requirements for HAL 0.5.10 "":

 - Linux kernel     >= 2.6.17
 - util-linux       >= 2.12r1    (--enable-umount-helper requires patch from RH #188193)
 - bash             >= 2.0
 - udev             >= 089
 - dbus             >= 0.61      (with glib bindings)
 - glib             >= 2.6.0
 - expat            >= 1.95.8
 - hal-info         >= 20070402
 - libusb           >= 0 0.1.10a (optional)
 - pciutils         >= 2.2.3     (optional
 - dmidecode        >= 2.7       (optional)
 - parted           == 1.7.1, 1.8.0, 1.8.1, 1.8.2, 1.86 (optional)
 - cryptsetup-luks  >= 1.0.1     (optional, needs LUKS patches)
 - libsmbios	    >= 0.13.4    (optional for DELL machines, Linux only)
 - dellWirelessCtl  >= 0.13.4    (optional, for Dell machines, must live in /usr/bin/, Linux only)
 - ConsoleKit       >= 0.2.0
 - pm-utils         >= 0.99.2 or newer (optional)


==========
HAL 0.5.9 "Precipice"
==========

Released April 2, 2007.

Requirements for HAL 0.5.9 "Precipice":

 - Linux kernel     >= 2.6.17
 - util-linux       >= 2.12r1    (--enable-umount-helper requires patch from RH #188193)
 - bash             >= 2.0
 - udev             >= 089
 - dbus             >= 0.61      (with glib bindings)
 - glib             >= 2.6.0
 - expat            >= 1.95.8
 - hal-info         >= 20070402
 - libusb           >= 0 0.1.10a (optional)
 - pciutils         >= 2.2.3     (optional
 - dmidecode        >= 2.7       (optional)
 - parted           == 1.7.1, 1.8.0, 1.8.1, 1.8.2, 1.86 (optional)
 - cryptsetup-luks  >= 1.0.1     (optional, needs LUKS patches)
 - libsmbios	    >= 0.13.4    (optional for DELL machines, Linux only)
 - dellWirelessCtl  >= 0.13.4    (optional, for Dell machines, must live in /usr/bin/, Linux only)
 - ConsoleKit       >= 0.2.0
 - pm-utils         >= 0.99.2 or newer (optional)

Contributors to HAL 0.5.9 "Precipice":

Alex Kanavin (1):
      provide textual interface description for USB devices

Andreas Hanke (1):
      fixes for AS_AC_EXPAND usage breaks hal-device-manager

Artem Kachitchkine (5):
      hald/linux/ids -> hald/ids and minor fixes
      fix minor memory leaks
      do not let tools coredump if HAL is not running
      Merge branch 'master' of ssh://git.freedesktop.org/git/hal
      solaris backend

Bastien Nocera (1):
      killswitch infrastructure for Bluetooth and use /dev/sonypi instead of spicctrl

Dan Nicholson (1):
      POSIX shells do not support '==' tests

Daniel Nylander (1):
      added Swedish translation file

Danny Kukawka (63):
      replaced LIBHAL_FREE_DBUS_ERROR with dbus_error_free()
      fixed acpi problems with incorrect low current voltage values
      fixed usage of dbus_error_is_set(error == NULL) due to D-Bus changes
      fixed detection of wireless devices with dscape kernel stack
      fix copy_property for merge in FDI files
      replace usage of g_assert() with error handling
      Merge branch 'master' of ssh://dkukawka@git.freedesktop.org/git/hal
      add hardware dependent/specific alsa sound devices to HAL
      fixed compiler warning
      fixed automake for linux
      fixed typo in 20-storage-methods.fdi
      remove unneeded close() calls
      fixed code docu in libhal
      removed not needed '.in' version of a luks script
      fixed segmentation fault in xen code
      Fix to guarantee that udi get generated before process FDI
      changed error handling for hal-storage-shared.c
      fixed detection of libsmbios to check for >= v0.11.6
      changed D-Bus error if /sbin/cryptsetup is not available
      remove sanitizing passwords for luks and prevent word splitting/path expansion
      fix detection of max. available omnibook brightness levels
      fix get/set brightness for sonypi/spiccrtl
      added new section for deprecated properties to spec
      fixed compiler warnings
      fix .hal-mtab mountpoint handling
      detect path to dmidecode on runtime (fd.o bug #9318)
      added hald-addon-dell-backlight to .gitignore
      fix gtk-doc code documentation
      fix (more) gtk-doc code documentation
      fixed typo in spec
      add copy_property FDI directive for stringlist properties
      fixed compiler warning in addon-dell-backlight
      fixed compiler warning
      fix compiler warning in libhal
      fixed compiler warning in create_cache.c
      fixed commit 882085a0, forgot cast to unsigned long
      typo fix
      fixed uid datatype in CallerInfo to unsigned long
      added new fdi match attribute 'contains_not' for strings and strlist
      added hald-probe-ieee1394-unit to .gitignore
      libhal: validate parameter to be != NULL
      fix expat detection in configure
      fixed code documentation
      added some files to .gitignore
      removed not needed string from translation files
      added support for ALSA MIDI devices
      validate error D-Bus names returned on stderror
      change usb*.version_bcd/usb*.speed_bcd to new double keys
      added hal-system-power-pm-is-supported to .gitignore
      fixed libsmbios search output to correct version
      updated requirements in NEWS file
      void hal-device-manager to open multiple about dialogs
      fix for new pciutils which need to link against zlib
      Merge branch 'master' of ssh://dkukawka@git.freedesktop.org/git/hal
      fix configure for DELL stuff
      fixed quoting and added allowed libparted versions to error output
      added .gitignore to new doc/man dir
      update several .gitignore files
      added optional support for IBM ACPI (hotkey) events.
      added configure option to enable build Toshiba ACPI button addon
      add more cdrom information
      update .gitignore
      updated German translation file

David Woodhouse (1):
      fix crash on PPC Pegasos that don't have PMU, ACPI or APM

David Zeuthen (196):
      post-release version bump to 0.5.9
      fix race condition when unmounting a volume takes a very long time

tools/hal-is-caller-locked-out.c
/***************************************************************************
 * CVSID: $Id$
 *
 * hal-is-caller-locked-out.c : Determine if a caller is locked out
 *
 * Copyright (C) 2007 David Zeuthen, <david@fubar.dk>
 *
 * Licensed under the Academic Free License version 2.1
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include <libhal.h>

/** 
 *  usage:
 *  @argc:                Number of arguments given to program
 *  @argv:                Arguments given to program
 *
 *  Print out program usage. 
 */
static void
usage (int argc, char *argv[])
{
	fprintf (stderr,
                 "\n"
                 "usage : hal-is-caller-locked-out --udi <udi> --interface <interface>\n"
                 "                                 --caller <caller-name>\n"
                 "                                 [--help] [--version]\n");
	fprintf (stderr,
                 "\n"
                 "        --udi            Unique Device Id\n"
                 "        --interface      Interface\n"
                 "        --caller         The name of the caller\n"
                 "        --version        Show version and exit\n"
                 "        --help           Show this information and exit\n"
                 "\n"
                 "This program determines if a given process on the system bus is\n"
                 "locked out of a D-Bus interface. If the process is locked out\n"
                 "or an error occurs this program exits with exit code 1. Otherwise\n"
                 "it exits with exit code 0. Note that only the super user (root)\n"
                 "or other privileged users can use this tool.\n"
                 "\n");
}

/** 
 *  main:
 *  @argc:                Number of arguments given to program
 *  @argv:                Arguments given to program
 *
 *  Returns:              Return code
 *
 *  Main entry point 
 */
int
main (int argc, char *argv[])
{
	char *udi = NULL;
	char *interface = NULL;
	char *caller = NULL;
        dbus_bool_t is_version = FALSE;
        dbus_bool_t locked_out;
	DBusError error;
        LibHalContext *hal_ctx;


	if (argc <= 1) {
		usage (argc, argv);
		return 1;
	}

	while (1) {
		int c;
		int option_index = 0;
		const char *opt;
		static struct option long_options[] = {
			{"udi", 1, NULL, 0},
			{"interface", 1, NULL, 0},
			{"caller", 1, NULL, 0},
			{"version", 0, NULL, 0},
			{"help", 0, NULL, 0},
			{NULL, 0, NULL, 0}
		};

		c = getopt_long (argc, argv, "",
				 long_options, &option_index);
		if (c == -1)
			break;

		switch (c) {
		case 0:
			opt = long_options[option_index].name;

			if (strcmp (opt, "help") == 0) {
				usage (argc, argv);
				return 0;
			} else if (strcmp (opt, "version") == 0) {
				is_version = TRUE;
			} else if (strcmp (opt, "udi") == 0) {
				udi = strdup (optarg);
			} else if (strcmp (opt, "caller") == 0) {
				caller = strdup (optarg);
			} else if (strcmp (opt, "interface") == 0) {
				interface = strdup (optarg);
			}
			break;

		default:
			usage (argc, argv);
			return 1;
			break;
		}
	}

	if (is_version) {
		printf ("hal-is-caller-locked-out " PACKAGE_VERSION "\n");
		return 0;
	}

	if (udi == NULL || caller == NULL || interface == NULL) {
		usage (argc, argv);
		return 1;
	}

	dbus_error_init (&error);	
	if ((hal_ctx = libhal_ctx_new ()) == NULL) {
		fprintf (stderr, "error: libhal_ctx_new\n");
		return 1;
	}
	if (!libhal_ctx_set_dbus_connection (hal_ctx, dbus_bus_get (DBUS_BUS_SYSTEM, &error))) {
		fprintf (stderr, "error: libhal_ctx_set_dbus_connection: %s: %s\n", error.name, error.message);
		LIBHAL_FREE_DBUS_ERROR (&error);
		return 1;
	}
	if (!libhal_ctx_init (hal_ctx, &error)) {
		if (dbus_error_is_set(&error)) {
			fprintf (stderr, "error: libhal_ctx_init: %s: %s\n", error.name, error.message);
			dbus_error_free (&error);
		}
		fprintf (stderr, "Could not initialise connection to hald.\n"
				 "Normally this means the HAL daemon (hald) is not running or not ready.\n");
		return 1;
	}

        locked_out = libhal_device_is_caller_locked_out (hal_ctx,
                                                         udi,
                                                         interface,
                                                         caller,
                                                         &error);
        if (dbus_error_is_set (&error)) {
		fprintf (stderr, "error: %s: %s\n", error.name, error.message);
		dbus_error_free (&error);
		return 1;
        }

        if (locked_out)
                return 1;
        else
                return 0;
}
